flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java

```diff
@@ -476,14 +482,13 @@ else if (numSources < parallelism) {
 	 * @return The preferred locations based in input streams, or an empty iterable,
 	 *         if there is no input-based preference.
 	 */
-	public Iterable<TaskManagerLocation> getPreferredLocationsBasedOnInputs() {
+	public Collection<CompletableFuture<TaskManagerLocation>> getPreferredLocationsBasedOnInputs() {
 		// otherwise, base the preferred locations on the input connections
 		if (inputEdges == null) {
 			return Collections.emptySet();
 		}
 		else {
-			Set<TaskManagerLocation> locations = new HashSet<>();
-			Set<TaskManagerLocation> inputLocations = new HashSet<>();
+			Set<CompletableFuture<TaskManagerLocation>> inputLocations = new HashSet<>(4);
```

Comment:
I think if we weigh inputs in scheduler maybe this problem can be solved. In above example, imagine that, a is "broadcast" and it downstream‘s parallelism is 10, and b is "forward" , then every ExecutionEdge for a's weight is ```a-total-weight / 10``` and ExecutionEdge of b's weight is ```b-total-weight / 1```, we can evaluate total weight according to vertex's throughput, currently, all vertex's total weight are equal, so location of b's input will be picked.

Reply:
This change was not entirely intended and the previous code makes totally sense with your explanation. I'm actually in favour of reverting my changes to not change the semantics for the moment.

However, for the future, I'm wondering whether this kind of decision should be made by the `ExecutionVertex` or whether it shouldn't rather be the task of the `Scheduler` to make this kind of decision. 

For example, what if a task has multiple input gates and one of them with exactly one producer. Then it will only return the location of this single producer. Now if this TM has no more slots left, then we would basically randomly pick another slot even though there might be other TMs one which another producer for this task would run.

Reply:
For example in the case of the broadcast join with lazy scheduling, it could be the case that the broadcasting operator produces first and thus triggers the `scheduleOrUpdateConsumers` call on the `ExecutionGraph`. This will then trigger the scheduling of the join operator. At this time, there might only be the location of the broadcast operator known. However, since we only return the forward operator's location future which has not been completed, the lazy scheduling will schedule without location preference because the `LocationPreferenceConstraint` is `ANY`.